### DISCUSSION TITLE
How to make objects available between modules

### DISCUSSION THREAD

**Discussion by zilch42** [zauberzeug/nicegui#1844] [Q&A] [answered]
### Question

I'm learning nicegui and struggling to get my head around scope and modularization. Apologies if this has been answered elsewhere, I've been looking at a lot of examples but may not be searching for the right things. The modularization example was helpful but doesn't pass any data between the modules.

I have an app with different pages in their own modules. I would like to load data on one page, and visualize it in another, with that data being user specific and persistent between changing pages.

Example
```python
# main.py
from nicegui import ui
import dataset
import plot 

@ui.page('/')
def main():
    ui.link("Table", "/table")
    ui.link("Plot", "/plot")
    
dataset.create()
plot.create()

ui.run()
```

```python
# dataset.py
from nicegui import ui
import pandas as pd
import numpy as np


def create():
    def create_data(): 
        dataset = pd.DataFrame(
            np.random.randint(0,100,size=(10, 2)), columns=list('XY'))
        datatable = ui.aggrid.from_pandas(df=dataset)

    @ui.page('/table')
    def page_dataset():
        ui.button("Create Data", on_click=lambda: create_data())
        ui.link("Home", "/")
```

```python
# plot.py
# this will 500 error because `dataset` isn't available
import plotly.graph_objects as go
from nicegui import ui
import pandas as pd
import numpy as np

# dataset = pd.DataFrame(
#             np.random.randint(0,100,size=(10, 2)), columns=list('XY'))

def create():
    @ui.page('/plot')
    def page_plot():
        if dataset is not None:
            fig = go.Figure(go.Scatter(x=dataset.X, y=dataset.Y))
            fig.update_layout(margin=dict(l=0, r=0, t=0, b=0))
            ui.plotly(fig).classes('w-full h-40')
        ui.link("Home", "/")
```

My main question is, how do I get `dataset` from `dataset.py` available in `plot.py`? (Currently whenever you click `Generate Data` you get an additional table which isn't what I want either, I just want one copy of `dataset`, but I think that's just a UI thing and hopefully it serves for example purposes.)

In my final version I will be loading quite a few objects from the dataset (models, dataframes, etc) and they won't be small so ideally I would like to be able to pass them around/refer to them as a set (a dict?) and not be making copies of them in memory.

I've come from R Shiny and the method there is to create a global reactiveValues object (dict basically) and pass that around between modules, but I'm not sure if that applies to nicegui. I haven't seen examples where pages have return values.

Thanks

**Comment by falkoschindler** [comment]
Hi @zilch42!

You can create the `dataset` globally in dataset.py and import it from plot.py:

dataset.py
```py
import numpy as np
import pandas as pd
from nicegui import ui

dataset = pd.DataFrame(np.random.randint(0, 100, size=(10, 2)), columns=list('XY'))

def create():
    @ui.page('/table')
    def page_dataset():
        ui.aggrid.from_pandas(df=dataset)
        ui.link("Home", "/")
```

plot.py
```py
import plotly.graph_objects as go
from dataset import dataset
from nicegui import ui

def create():
    @ui.page('/plot')
    def page_plot():
        ui.plotly(go.Figure(go.Scatter(x=dataset.X, y=dataset.Y)))
        ui.link("Home", "/")
```

Is that what you're looking for?

**Comment by zilch42** [comment]
Thanks @falkoschindler, that gets me a bit closer, and works ok if the `dataset` is static, but I haven't got it to work if the `dataset` is produced dynamically by some action in the ui (in my example, from clicking a button; in my final app, loaded from a folder that the user selected from the file system). How can I get `plot.py` to be able to see the change performed on the dataset page? (Sorry if my example was too simple, I was hoping the dependency on the button would be enough). 

I've modified my example using a refreshable `data_table_ui` object and importing it on `plot.py`, which gets me closer (I no longer have multiple tables after multiple clicks, and the table is persistent on the dataset page after navigating home and then back to it, and the plot page doesn't throw an error, but it still can't see the change in `dataset` when the button gets clicked. 

```python
# main.py
from nicegui import ui
import dataset
import plot 

@ui.page('/')
def main():
    ui.link("Table", "/table")
    ui.link("Plot", "/plot")
    
dataset.create()
plot.create()

ui.run()
```

```python
# dataset.py
from nicegui import ui
import pandas as pd
import numpy as np

dataset = None

@ui.refreshable
def data_table_ui():
    if dataset is not None:
        ui.aggrid.from_pandas(df=dataset)

def create():
    def create_data(): 
        global dataset
        dataset = pd.DataFrame(
            np.random.randint(0,100,size=(10, 2)), columns=list('XY'))
        data_table_ui.refresh()
        
    @ui.page('/table')
    def page_dataset():
        ui.button("Create Data", on_click=lambda: create_data())
        data_table_ui()
        ui.link("Home", "/")
```

```python
# plot.py
import plotly.graph_objects as go
from nicegui import ui
import pandas as pd
import numpy as np
from dataset import dataset

@ui.refreshable
def plot_ui():
    global dataset 
    if dataset is not None:
        fig = go.Figure(go.Scatter(x=dataset.X, y=dataset.Y))
        fig.update_layout(margin=dict(l=0, r=0, t=0, b=0))
        ui.plotly(fig).classes('w-full h-40')

def create():
    @ui.page('/plot')
    def page_plot():
        plot_ui()
        ui.link("Home", "/")
```

**Reply by falkoschindler** [reply]
Ok, one problem is that you import `from dataset import dataset`, which is `None` at this time. As soon as the object `dataset` is updated, the imported reference is outdated (still `None`). You can simply `import dataset` instead and use the object like `dataset.dataset` (you could use a less confusing name, e.g. `data.df` to better distinguish the object `df` from the module `data`). This will update the plot as expected _when then browser tab is reloaded_.

This leaves us with the problem of how to notify the plot module to update the plotly figure without reloading the tab. There are probably many ways to do that, but you basically have to decide, which module should depend on which other module. In the following example I'll make plot.py independent of data.py and let data.py drive all UI updates. Therefore I'll change the refreshable to accept the dataframe as an argument:

```py
# main.py
import data
import plot
from nicegui import ui

@ui.page('/')
def main():
    ui.link("Table", "/table")
    ui.link("Plot", "/plot")

data.create()
plot.create(data.df)

ui.run()
```

```py
# data.py
import numpy as np
import pandas as pd
import plot
from nicegui import ui

df = None

@ui.refreshable
def data_table_ui():
    if df is not None:
        ui.aggrid.from_pandas(df=df)

def create():
    def create_data():
        global df
        df = pd.DataFrame(np.random.randint(0, 100, size=(10, 2)), columns=list('XY'))
        data_table_ui.refresh()
        plot.plot_ui.refresh(df)

    @ui.page('/table')
    def page_table():
        ui.button("Create Data", on_click=create_data)
        data_table_ui()
        ui.link("Home", "/")
```

```py
# plot.py
import pandas as pd
import plotly.graph_objects as go
from nicegui import ui

@ui.refreshable
def plot_ui(df: pd.DataFrame | None):
    if df is not None:
        fig = go.Figure(go.Scatter(x=df.X, y=df.Y))
        fig.update_layout(margin=dict(l=0, r=0, t=0, b=0))
        ui.plotly(fig).classes('w-full h-40')

def create(df: pd.DataFrame | None):
    @ui.page('/plot')
    def page_plot():
        plot_ui(df)
        ui.link("Home", "/")
```

As I said, there might be other ways to do it. Even an event system could useful, but requires additional infrastructure.

**Reply by zilch42** [reply, marked as answer]
Thanks @falkoschindler, that's great. Gets me closer. If I have 2 tabs open at the at same time I can see the plot update on one tab based on changes to the dataset in the other tab, but no plot if I navigate from data > home > plot in a single tab. 

I played with it some more and got it working in a single tab by defining a class for my data in `main` with a `refresh` method and passing it to both pages. See below for anyone who else who finds this 

```python
# main.py
import data
import plot
from nicegui import ui

class refreshable_df_class():
    def __init__(self):
        self.df = None
    def refresh(self, df):
        self.df = df
        
data_model = refreshable_df_class()

@ui.page('/')
def main():
    ui.link("Table", "/table")
    ui.link("Plot", "/plot")

data.create(data_model)
plot.create(data_model)

ui.run()
```

```python
# data.py
import numpy as np
import pandas as pd
from nicegui import ui

df = None

@ui.refreshable
def data_table_ui():
    if df is not None:
        ui.aggrid.from_pandas(df=df)

def create(data_model):
    def create_data():
        global df
        df = pd.DataFrame(np.random.randint(0, 100, size=(10, 2)), columns=list('XY'))
        data_model.refresh(df)
        data_table_ui.refresh()
        # plot.plot_ui.refresh(df)

    @ui.page('/table')
    def page_table():
        ui.button("Create Data", on_click=create_data)
        data_table_ui()
        ui.link("Home", "/")
```

```python
# plot.py
import pandas as pd
import plotly.graph_objects as go
from nicegui import ui
from typing import Union

@ui.refreshable
def plot_ui(data_model: Union[pd.DataFrame, None]):
    if data_model.df is not None:
        fig = go.Figure(go.Scatter(x=data_model.df.X, y=data_model.df.Y))
        fig.update_layout(margin=dict(l=0, r=0, t=0, b=0))
        ui.plotly(fig).classes('w-full h-40')

def create(data_model: Union[pd.DataFrame, None]):
    @ui.page('/plot')
    def page_plot():
        plot_ui(data_model)
        ui.link("Home", "/")
```

Thanks for all your help with this @falkoschindler, I would not have got there without you!
